README.md
# SDK-RaaS-iOS

**The following guide explains how to install and implement Pangea RaaS' Mobile SDK into your application environment.** 
- Version number: 1.0.4

## [Changelog](https://github.com/gopangea/sdk-raas-ios/blob/master/CHANGELOG.md)

### Introduction


The Mobile SDK is analogous to [Pangea's Javascript](https://connect-raas-api.pangeamoneytransfer.com/?java#pangea-js-library) code used for fraud review enhancement on the browser and Funding card tokenization.

Pangea requires the use of our SDK to mitigate fraud and to securely collect debit card information for payment. 
Include this library in your ios application using the following instructions.


**To be able to use Pangea RaaS SDK in your application, Pangea must add your hostname to the CORS whitelist.**


### Prerequisites:

iOS 10.0 and above.

### Installation:

**Using CocoaPods:**
  * Add the following line to the Podfile, under your app's target section:
```
pod 'RaaS_SDK', :git => 'https://github.com/gopangea/sdk-raas-ios.git', :tag => 'Version number'
```
  * Run $ pod install in your project directory.
  * Open App.xcworkspace and build.
 
### Implementation:


1. On the top of your ApplicationDelegate add 
```
import RaaS_SDK
```
2. Create the session at the end of applicationDidFinishLaunching:
```
Pangea.sharedInstance.createSession( pangeaSessionID: "NumberSessionHere", debugInfo: true, environment: .INTEGRATION)
```

* createSession() arguments:
  - pangeaSessionID - Unique String identifier of the user's current session in the app. The same identifier is passed by your backend to Pangea.
  - debugInfo - Boolean parameter that enables pangea logging for debugging.
  - environment - Use it to point to the desired environment, there are three: *PRODUCTION*, *DEV* and *INTEGRATION*

  
2. if you need to retrive the session from Pangea instance use the following function:
```
Pangea.sharedInstance.getSessionId()
```

3. When the user's session changes, update the token by calling:
```
Pangea.sharedInstance.updateSessionToken("newToken");
```
4. Notify to pangea on each relevant HTTP request from the app:
```
Pangea.sharedInstance.logRequest("https://<request URL>");
```
5. Collect rich device information. This method should usually be called only once per session:
```
Pangea.sharedInstance.logSensitiveDeviceInfo();
```


### Funding card tokenization:

Pangea RaaS iOS sdk must also be used for generating a temporary token representing a sender’s funding card. Pangea retrieves the token by first encrypting the card data using your public key then sends it to Pangea’s server. You will send the temporary token to your server and finally over to the Pangea RaaS API to create a funding card. Pangea uses the temporary token to generate a permanent token with our card payments provider. The card number and cvv is never stored anywhere in Pangea. You will use the same permanent token for each transfer made with that card. The temporary token is valid for 24 hours. Upon successful generation of the temporary token, you can safely send it to your server. Then to create the funding method in the RaaS API, provide the token along with the other required fields: 

```
Pangea.sharedInstance.createToken(cardInfo: CardInformation, completion: @escaping QueryResult<TokenResponse>) 
```
**The structure of Card information is:**

```
struct CardInformation {
    var publicKey: String
    var partnerIdentifier: String
    var cardNumber: String
    var cvv: String
}
```
* CardInformation() arguments:
  - publicKey - Unique RAS pubic key identifier of provided by Pangea, pass this argument as a String with the key only without any blank/enter/tab/etc space, a valid string could be: 
  
  ```
  "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAq7BL3rzZbWx6xDmxDxozfUhoJ2xJawfKoGqBgqUa+ZTWUYUtkrCMuS3l8bKZZij4MQQmFb4vvIUJ0AoY0aVK59uxom1MEA9X89Vaz0Ctv5TNdjm7NQN3oosdtKeMd7g1fAxBXoR2XdShM9Nq0IjNHgWbbgFlq4CTKdPyG7N/M5eAnSjDOO9xIADZ9DsWGk3TgZGKbr36EJGYfT8R1E/l+/2YRLVlKf/lLGkl0LSPJ+kv4icB7i48v2GTTAyRs04oFPc9xB/JdoCxCtUmaIcy
  vsjavj9MxRZ3ubOFLNdh8SJ3GmVgRMndxvJGKAVAeURP4eGFK9btnLan9Kzt6BXcFQIDAQAB"
  ```

  - partnerIdentifier -  This is the name of the shop
  - cardNumber - A valid card number, this number will be encryped and send to the pangea server
  - cvv -  A cvv number, this number will be encryped and send to the pangea server
  
You’ll have a different public key for both sandbox and production. The partnerIdentifier will be the same in both environments and will be assigned to you.

**QueryResult Closure**

This closure will let you retrieve the token from pagea when the service call is completed, it has two parameters
```
typealias QueryResult<T> = (T?, String?) -> Void
```

You can use something similar to this:
  ```
Pangea.sharedInstance.createToken(cardInfo: card, completion: {tokenResponse,error in
                     if let token:TokenResponse = tokenResponse
                     {
                        //do something with your Token
                        print(token.token)
                     }
                     else {
                        // failed operation
                        print(error)
                     }
                     })
```
TokenResponse is just a wrapper for a String

**Client Session Data**

From version 1.0.4 we introduced a new method to get a base64 encoded string which the RaaS partner will store in their database instead of the clientSessionId. This encoded string is a JSON object containing some platform RaaS identifiers and the client session id provided by your implementation.

```
Pangea.sharedInstance.getClientData(completion: @escaping QueryResult<String>) 
```

This id is the same that the sessionID, if you already provided one is not necessary to provide another before calling this method, (when you create your pangea instance you pass this ID as the parameter pangeaSessionID)


You can use something similar to this to retrive your encoded client session data:

```
        Pangea.sharedInstance.getClientData(completion: {clientInfo,error in
                         if let clientInfo:String = clientInfo
                         {
                            //do something with your data encoded string 
                            print(clientInfo)
                         }
                         else {
                            // failed operation
                            print(error)
                         }
                         })
```



### FAQs

Q: Does the SDK require any external frameworks or additional permissions from the user?
* The SDK does not require or prompt for any user permissions.
* The SDK utilizes certain permissions if the host app previously requested them.

Q: When and how should logRequest() be used?
* Call this method every time a user performs a meaningful action inside the flow of the remittance (for
example: start a new remittance , search a contact, finalize the remittance).
* Pangea uses this data for behavioral modeling and analysis, so please take care
to only call logRequest() in response to an actual user action.
* The single URL argument represents the action taken, this can usually be
achieved by passing the same URL used in the backend call for the action (ie.,
https://api.myshop.com/remittance/).

Q: When should logsensitiveDeviceInfo() be called?
* Ideally once per session, depending on the use case. For example - before
attempting to process a remmitance.
* Multiple invocations of this function in a single session will not generate an error,
but have the wasteful effect of generating identical payloads for analysis.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- No changes comming

## [1.0.4] - 2021-02-25
### Added
- A new method that will return a base64 encoded string which the RaaS partner will store in their database instead of the clientSessionId
```
getClientData()
```
### Changed
- The QueryResull has chaged to a generic type, in this way can it be use to wrap any type of reponse from the pagea sdk that could be added in the future, the refactored method is this one:

```
typealias QueryResult<T> = (T?, String?) -> Void
```
You can use it as the same as the previous one:
```
Pangea.sharedInstance.createToken(cardInfo: card, completion: {tokenResponse,error in
     guard let token = tokenResponse else { //invalid token }
})
        
Pangea.sharedInstance.getClientData(completion: {clientInfo,error in
     guard let clientData = clientInfo else { //invalid clientInfo }
})
```

## [1.0.3] - 2020-12-08
### Added
- A custom header x-pangea-user-agent that describes the version number, build number and OS 
### Changed
- The debug mode now shows the body, url and headers from the Http request

## [0.0.9] - 2020-09-29
### Changed
- Initial version of the SDK


RaaS_SDK/Models.swift
//
//  Models.swift
//  RaaSCardTokenizerIOSExample
//
//  Created by Carlos Hernandez on 22/09/20.
//  Copyright © 2020 Pangea. All rights reserved.
//

import Foundation

public struct CardInformation {
    var publicKey: String
    var partnerIdentifier: String
    var cardNumber: String
    var cvv: String
    public init(
        publicKey: String,
        partnerIdentifier: String,
        cardNumber: String,
        cvv: String){
        self.publicKey = publicKey
        self.partnerIdentifier =  partnerIdentifier
        self.cardNumber  =  cardNumber
        self.cvv =  cvv
    }
}

struct TokenRequest {
    var encryptedCardNumber: String = ""
    var encryptedCvv: String = ""
    var partnerIdentifier: String = ""
    var requestId: String = ""
}

public struct TokenResponse {
    public var token: String
}

internal func getRequestId() -> String{
    let randomNumber = String(drand48()).replacingOccurrences(of:".",with: "")
    let randomNumberString = (randomNumber as NSString).longLongValue
    let requestId = String(randomNumberString, radix: 11)
    return requestId
}
public typealias QueryResult<T> = (T?, String?) -> Void // this is for the client user
internal typealias QueryData = (Data?, Error?) -> Void //this is just for the internal query to the network

public enum Environment: String {
    case PRODUCTION
    case DEV
    case INTEGRATION
}

internal func getVersion()->String{
    return "1.0.4"
}

internal func getBuild()->String{
    return "6"
}

internal struct MyError: Error {
    let msg: String

}
extension MyError: LocalizedError {
    public var errorDescription: String? {
            return NSLocalizedString(msg, comment: "")
    }
}


RaaS_SDK/Pangea.swift
//
//  Pangea.swift
//  RaaS_SDK
//
//  Created by Carlos Hernandez on 22/09/20.
//  Copyright © 2020 Pangea. All rights reserved.
//

import Foundation



import Foundation
public class Pangea {
    private static let TAG = "Pangea"
    private var debugInfo : Bool = false
    private var pangeaSessionID :String = ""
    public static let sharedInstance = Pangea()
    private (set) var api = "https:api.pangea-raas-integration.com/raas/v1/tokenization/card"
    private var environment:Environment = .DEV
    
    private init() {
    }
    
    public func createSession(pangeaSessionID: String,debugInfo: Bool, environment:Environment) {
        RiskifiedBeacon.start("gopangea.com", sessionToken:pangeaSessionID, debugInfo: debugInfo)
        self.debugInfo = debugInfo
        self.pangeaSessionID=pangeaSessionID
        self.environment = environment
    }
    
    public func getSessionId() -> String{
        return self.pangeaSessionID
    }
    
    
    public func updateSessionToken(newPangeaSessionID: String) {
        self.pangeaSessionID = newPangeaSessionID
        RiskifiedBeacon.updateSessionToken(newPangeaSessionID)
    }
    
    public func logRequest(url:URL) {
        RiskifiedBeacon.logRequest(url)
    }
    
    public func logSensitiveDeviceInfo() {
        RiskifiedBeacon.logSensitiveDeviceInfo()
    }
    
    public func createToken(cardInfo: CardInformation, completion: @escaping QueryResult<TokenResponse>) {
        let completionAxiliary:QueryData = {data,error in
            if let error = error {
                let errorMessage = "DataTask error: " + error.localizedDescription + "\n"
                completion(nil,errorMessage)
            } else if let data = data{
                let respondMessage = String(decoding: data, as: UTF8.self)
                let tokenResponse = TokenResponse(token:respondMessage)
                if (self.debugInfo) {
                    print(tokenResponse.token)
                }
                completion(tokenResponse,nil)
            }
        }
        RaaSService(environment: environment, debugInfo:debugInfo).createToken(cardInfo: cardInfo, completion: completionAxiliary)
    }
    
    
    
    public func getClientData(completion: @escaping QueryResult<String>) {
        let completionAxiliary:QueryData = {data,error in
            if let error = error {
                let errorMessage = "DataTask error: " + error.localizedDescription + "\n"
                completion(nil,errorMessage)
            } else if let data = data{
                do {
                    var json = try JSONSerialization.jsonObject(with: data, options:.mutableContainers) as? [String : Any]
                    json?.merge(["clientSessionId":self.getSessionId()]) {(current,_) in current}
                    if let jsonString = jsonToString(json: json as Any, isDebug: self.debugInfo){
                        let trimmedJsonString = self.getEncodedInfoUser(jsonUserInfo: jsonString)
                        completion(trimmedJsonString, nil)
                    }else{
                        completion(nil, "error when covert json response to string , JSON =  \(String(describing: json?.description))")
                    }
                } catch let error {
                    completion(nil, error.localizedDescription)
                }
            }
        }
        RaaSService(environment: environment, debugInfo:debugInfo).fetchClientInfo(completion: completionAxiliary)
    }
    
    private func getEncodedInfoUser(jsonUserInfo: String) -> String?{
        let trimmedJsonString = jsonUserInfo.replacingOccurrences(of: "\n", with: "").replacingOccurrences(of: " ", with: "")
        return trimmedJsonString.base64Encoded()
    }
    
}



RaaS_SDK/RSAUtil.swift
//
//  RSAUtil.swift
//  RaaSCardTokenizerIOSExample
//
//  Created by Carlos Hernandez on 17/09/20.
//  Copyright © 2020 Pangea. All rights reserved.
//

import Foundation

internal class RSAUtil {
    static func encrypt(string: String, publicKey: String?) -> String? {
        guard let publicKey = publicKey else { return nil }

        let keyString = publicKey.replacingOccurrences(of: "-----BEGIN PUBLIC KEY-----", with: "").replacingOccurrences(of: "-----END PUBLIC KEY-----", with: "")
        guard let data = Data(base64Encoded: keyString) else { return nil }

        var attributes: CFDictionary {
            return [kSecAttrKeyType         : kSecAttrKeyTypeRSA,
                    kSecAttrKeyClass        : kSecAttrKeyClassPublic,
                    kSecAttrKeySizeInBits   : 2048,
                    kSecReturnPersistentRef : kCFBooleanTrue!] as CFDictionary
        }

        var error: Unmanaged<CFError>? = nil
        guard let secKey = SecKeyCreateWithData(data as CFData, attributes, &error) else {
            print(error.debugDescription)
            return nil
        }
        return _encrypt(string: string, publicKey: secKey)
    }

    static private func _encrypt(string: String, publicKey: SecKey) -> String? {
        let buffer = [UInt8](string.utf8)

        var keySize   = SecKeyGetBlockSize(publicKey)
        var keyBuffer = [UInt8](repeating: 0, count: keySize)

        // Encrypto  should less than key length
        guard SecKeyEncrypt(publicKey, SecPadding.PKCS1, buffer, buffer.count, &keyBuffer, &keySize) == errSecSuccess else { return nil }
        return Data(bytes: keyBuffer, count: keySize).base64EncodedString()
    }
}

extension String {
    //"heroes".base64Encoded() // It will return: aGVyb2Vz
    //"aGVyb2Vz".base64Decoded() // It will return: heroes
    
    func base64Encoded() -> String? {
        return data(using: .utf8)?.base64EncodedString()
    }
    func base64Decoded() -> String? {
        guard let data = Data(base64Encoded: self) else { return nil }
        return String(data: data, encoding: .utf8)
    }
}

internal func jsonToString(json: Any, isDebug: Bool) -> String?{
    do {
        // first of all convert json to the data
        let data1 =  try JSONSerialization.data(withJSONObject: json, options: JSONSerialization.WritingOptions.prettyPrinted)
        // the data will be converted to the string
        let convertedString = String(data: data1, encoding: String.Encoding.utf8)
        return convertedString
    } catch let myJSONError {
        if isDebug {
            print(myJSONError)
        }
        return nil
    }
    
}




 


RaaS_SDK/CardUtils.swift
//
//  CardUtils.swift
//  RaaSCardTokenizerIOSExample
//
//  Created by Carlos Hernandez on 22/09/20.
//  Copyright © 2020 Pangea. All rights reserved.
//

import Foundation


internal func isCardNumberValid(cardNumber: String)->Bool{
    var card = cardNumber
    card.removingRegexMatches(pattern: "\\D+", replaceWith: "")
    var numbers = card.map { character in
        Int(String(character))
    }
    numbers.reverse()
    if (numbers.count < 12) {
        return false
    }
    var total = 0
    for (index, element) in numbers.enumerated() {
        if element != nil{
            if (index % 2 != 0){
                if (numbers[index]! > 4){
                    total += 2 * numbers[index]! - 9
                }else{
                    total += 2 * numbers[index]! - 0
                }
            }else{
                 total += numbers[index]!
            }
        }else{
            //conversion error, this cannot happen because we are filtering  only numbers and the conversion should happen all the times
            return false
        }
    }
    return total % 10 == 0
}

// http://www.ietf.org/rfc/rfc4122.txt
internal func createUUID()-> String {
    var s = [Character](repeating: " ", count: 36)
    let hexDigits = Array("0123456789abcdef")
    for (index, _) in s.enumerated() {
        let r = Int(arc4random_uniform(16)) ;
        s[index] = hexDigits[r]
       }
    s[4] = "4"
    var temporalNumber = s[19].hexDigitValue
    if containsMatch(of: "\\D+", inString: String(s[19])) {
        temporalNumber = 0x0
    }
    temporalNumber = ((temporalNumber! & 3)|8)
    s[19] = String(temporalNumber!, radix: 16).first! //s[19] can be only 8,9,a or b
    s[8] = "-"; s[13] = "-";s[18] = "-"; s[23] = "-"
    return String(s)
}

internal extension String {
    mutating func removingRegexMatches(pattern: String, replaceWith: String = "") {
        do {
            let regex = try NSRegularExpression(pattern: pattern, options: NSRegularExpression.Options.caseInsensitive)
            let range = NSMakeRange(0, self.count)
            self = regex.stringByReplacingMatches(in: self, options: [], range: range, withTemplate: replaceWith)
        } catch {
            return
        }
    }
}



internal func containsMatch(of pattern: String, inString string: String) -> Bool {
  guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
    return false
  }
  let range = NSRange(string.startIndex..., in: string)
  return regex.firstMatch(in: string, options: [], range: range) != nil
}


RaaS_SDK/TokenService.swift
//
//  TokenService.swift
//  RaaSCardTokenizerIOSExample
//
//  Created by Carlos Hernandez on 17/09/20.
//  Copyright © 2020 Pangea. All rights reserved.
//

import Foundation


import Foundation


//MARK: - TokenService


// Runs query data task only
internal class TokenService {
    private let environment :Environment
    private var debugInfo:Bool
    public init(environment: Environment, debugInfo:Bool){
        self.environment = environment
        self.debugInfo = debugInfo
    }
    
    //MARK: - Constants
    private let defaultSession = URLSession(configuration: .default)
    
    //MARK: - Variables And Properties
    
    private var getTokenTask: URLSessionDataTask?
    private var getClientInfoTask: URLSessionDataTask?
    
    //MARK: - Token
    //returns a pair value, if the secon value has an empty screen the response of the fuction is
    //sucessfull but if tot it will return an error message
    private func getTokenAPIBodyObject(cardInfo: CardInformation) -> (body:[String:String], errorMessage:String){
        if !isCardNumberValid(cardNumber: cardInfo.cardNumber){
            return (["":""], "invalid Card Number: \(cardInfo.cardNumber)")
        }
        guard let encryptedCardNumber = RSAUtil.encrypt(string: cardInfo.cardNumber, publicKey: cardInfo.publicKey) else {
            return (["":""], "encryption for card number Failed")
        }
        guard let encryptedCvv = RSAUtil.encrypt(string: cardInfo.cvv, publicKey: cardInfo.publicKey) else {
            return (["":""], "encryption for cvv Failed")
        }
        let body = (["requestId":getRequestId(),
                    "partnerIdentifier":cardInfo.partnerIdentifier,
                    "encryptedCardNumber":encryptedCardNumber,
                    "encryptedCvv":encryptedCvv], "")
        return body
    }
    
    
 
    internal func createToken(cardInfo: CardInformation, completion: @escaping QueryData) {
        getTokenTask?.cancel()
        defer {getTokenTask = nil}
        let bodyObject = getTokenAPIBodyObject(cardInfo: cardInfo)
        if bodyObject.errorMessage != "" {
            completion(nil, MyError(msg: bodyObject.errorMessage))
            return
        }
        let body = bodyObject.body
        if let request = getRequest(path: "v1/tokenization/card", body: body,httpMethod: "POST") {
            getTokenTask = getDataTask(request: request, completion: completion, debugInfo: debugInfo)
            getTokenTask?.resume()
        }else {
            completion(nil,MyError(msg: "error creating the request"))
        }
    }
    
    //    MARK: - Client info
    
    internal func fetchClientInfo(completion: @escaping QueryData) {
        getClientInfoTask?.cancel()
        defer {getClientInfoTask = nil}
        let emptyBody : [String: String]? = nil
        if let request = getRequest(path: "v1/clientSession/data", body: emptyBody,httpMethod: "GET") {
            getClientInfoTask = getDataTask(request: request, completion: completion, debugInfo: debugInfo)
            getClientInfoTask?.resume()
        }else {
            completion(nil,MyError(msg: "error creating the request"))
        }
    }
    
    //    MARK: - General Methods
    private func getDataTask(request: URLRequest, completion: @escaping QueryData, debugInfo: Bool) -> URLSessionDataTask {
        return defaultSession.dataTask(with: request) {data, response, error in
            if debugInfo{
                print("RESPONSE INFO:")
                print("response:  \(String(describing: response))")
                if let data = data{
                    print("Body/data: \(String(decoding: data , as: UTF8.self))")
                }
                print("error: \(String(describing: error))")
            }
            if let error = error {
                DispatchQueue.main.async {
                    completion(nil,error)
                }
            } else if let data = data{
                DispatchQueue.main.async {
                    completion(data,nil)
                }
            }
        }
    }
    //retuns the simplest base url with the minimn headers required, you have to pass the verb method
    //like "POST", or "GET"
    //path: is the next segment of the url where the resource is located
    //if request doesn't have body pass a nil
    internal func getRequest<T>(path: String,body:T?, httpMethod: String) -> URLRequest?{
        let fullUrl =  ("\(getApi(environment: environment))\(path)")
        if let urlComponents = URLComponents(string: fullUrl) {
            guard let url = urlComponents.url else {
                if(debugInfo){
                    print("url malformed \(String(describing: urlComponents.url))")
                }
                return nil
            }
            let versionNumber = getVersion()
            let buildNumber = getBuild()
            let userAgent = "RaasSdk|iOS|Version=\(versionNumber)|Build=\(buildNumber)"
            var request = URLRequest(url: url)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type") //headers
            request.setValue(userAgent, forHTTPHeaderField: "x-pangea-user-agent")
            request.httpMethod = httpMethod//verb
            var bodyString : String? = nil
            if body != nil {
                let bodyData = try? JSONSerialization.data(withJSONObject: body!)
                request.httpBody = bodyData //adding data
                bodyString = String(decoding: request.httpBody! , as: UTF8.self)
            }
            if(debugInfo){
                print("\(request.httpMethod ?? "") \(String(describing: request.url) )")
                print("BODY \n \(bodyString ?? "")")
                print("HEADERS \n \(String(describing: request.allHTTPHeaderFields) )")
            }
            return request
        }
        return nil
    }
    
    private func getApi(environment: Environment) ->String {
        var api = "noUrlFound"
        switch environment {
        case .PRODUCTION:
            api = "https://api.pangea-raas.com/raas/"
        case .DEV:
            api = "https://api.pangea-raas-dev.com/raas/"
        case .INTEGRATION:
            api = "https:api.pangea-raas-integration.com/raas/"
        }
        return api
    }
    
    
}



RaaS_SDK/RaaSService.swift
//
//  TokenService.swift
//  RaaSCardTokenizerIOSExample
//
//  Created by Carlos Hernandez on 17/09/20.
//  Copyright © 2020 Pangea. All rights reserved.
//

import Foundation


import Foundation


//MARK: - RasSService


// Runs query data task only
internal class RaaSService {
    private let environment :Environment
    private var debugInfo:Bool
    public init(environment: Environment, debugInfo:Bool){
        self.environment = environment
        self.debugInfo = debugInfo
    }
    
    //MARK: - Constants
    private let defaultSession = URLSession(configuration: .default)
    
    //MARK: - Variables And Properties
    
    private var getTokenTask: URLSessionDataTask?
    private var getClientInfoTask: URLSessionDataTask?
    
    //MARK: - Token
    //returns a pair value, if the secon value has an empty screen the response of the fuction is
    //sucessfull but if tot it will return an error message
    private func getTokenAPIBodyObject(cardInfo: CardInformation) -> (body:[String:String], errorMessage:String){
        if !isCardNumberValid(cardNumber: cardInfo.cardNumber){
            return (["":""], "invalid Card Number: \(cardInfo.cardNumber)")
        }
        guard let encryptedCardNumber = RSAUtil.encrypt(string: cardInfo.cardNumber, publicKey: cardInfo.publicKey) else {
            return (["":""], "encryption for card number Failed")
        }
        guard let encryptedCvv = RSAUtil.encrypt(string: cardInfo.cvv, publicKey: cardInfo.publicKey) else {
            return (["":""], "encryption for cvv Failed")
        }
        let body = (["requestId":getRequestId(),
                    "partnerIdentifier":cardInfo.partnerIdentifier,
                    "encryptedCardNumber":encryptedCardNumber,
                    "encryptedCvv":encryptedCvv], "")
        return body
    }
    
    
 
    internal func createToken(cardInfo: CardInformation, completion: @escaping QueryData) {
        getTokenTask?.cancel()
        defer {getTokenTask = nil}
        let bodyObject = getTokenAPIBodyObject(cardInfo: cardInfo)
        if bodyObject.errorMessage != "" {
            completion(nil, MyError(msg: bodyObject.errorMessage))
            return
        }
        let body = bodyObject.body
        if let request = getRequest(path: "v1/tokenization/card", body: body,httpMethod: "POST") {
            getTokenTask = getDataTask(request: request, completion: completion, debugInfo: debugInfo)
            getTokenTask?.resume()
        }else {
            completion(nil,MyError(msg: "error creating the request"))
        }
    }
    
    //    MARK: - Client info
    
    internal func fetchClientInfo(completion: @escaping QueryData) {
        getClientInfoTask?.cancel()
        defer {getClientInfoTask = nil}
        let emptyBody : [String: String]? = nil
        if let request = getRequest(path: "v1/clientSession/data", body: emptyBody,httpMethod: "GET") {
            getClientInfoTask = getDataTask(request: request, completion: completion, debugInfo: debugInfo)
            getClientInfoTask?.resume()
        }else {
            completion(nil,MyError(msg: "error creating the request"))
        }
    }
    
    //    MARK: - General Methods
    private func getDataTask(request: URLRequest, completion: @escaping QueryData, debugInfo: Bool) -> URLSessionDataTask {
        return defaultSession.dataTask(with: request) {data, response, error in
            if debugInfo{
                print("RESPONSE INFO:")
                print("response:  \(String(describing: response))")
                if let data = data{
                    print("Body/data: \(String(decoding: data , as: UTF8.self))")
                }else{
                    print("error: \(String(describing: error))")
                }
            }
            if let error = error {
                DispatchQueue.main.async {
                    completion(nil,error)
                }
            } else if let data = data{
                DispatchQueue.main.async {
                    completion(data,nil)
                }
            }
        }
    }
    //retuns the simplest base url with the minimn headers required, you have to pass the verb method
    //like "POST", or "GET"
    //path: is the next segment of the url where the resource is located
    //if request doesn't have body pass a nil
    internal func getRequest<T>(path: String,body:T?, httpMethod: String) -> URLRequest?{
        let fullUrl =  ("\(getApi(environment: environment))\(path)")
        if let urlComponents = URLComponents(string: fullUrl) {
            guard let url = urlComponents.url else {
                if(debugInfo){
                    print("url malformed \(String(describing: urlComponents.url))")
                }
                return nil
            }
            let versionNumber = getVersion()
            let buildNumber = getBuild()
            let userAgent = "RaasSdk|iOS|Version=\(versionNumber)|Build=\(buildNumber)"
            var request = URLRequest(url: url)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type") //headers
            request.setValue(userAgent, forHTTPHeaderField: "x-pangea-user-agent")
            request.httpMethod = httpMethod//verb
            var bodyString : String? = nil
            if body != nil {
                let bodyData = try? JSONSerialization.data(withJSONObject: body!)
                request.httpBody = bodyData //adding data
                bodyString = String(decoding: request.httpBody! , as: UTF8.self)
            }
            if(debugInfo){
                print("\(request.httpMethod ?? "") \(String(describing: request.url) )")
                print("BODY \n \(bodyString ?? "")")
                print("HEADERS \n \(String(describing: request.allHTTPHeaderFields) )")
            }
            return request
        }
        return nil
    }
    
    private func getApi(environment: Environment) ->String {
        var api = "environmentNotFound"
        switch environment {
        case .PRODUCTION:
            api = "https://api.pangea-raas.com/raas/"
        case .DEV:
            api = "https://api.pangea-raas-dev.com/raas/"
        case .INTEGRATION:
            api = "https:api.pangea-raas-integration.com/raas/"
        }
        return api
    }
    
    
}

